alacritty/src/display.rs

```diff
@@ -391,45 +389,34 @@ impl Display {
         let cell_width = self.size_info.cell_width;
         let cell_height = self.size_info.cell_height;
 
-        // Recalculate padding.
-        let padding = config.ui_config.window.padding;
-        let mut padding_x = f32::from(padding.x) * self.size_info.dpr as f32;
-        let mut padding_y = f32::from(padding.y) * self.size_info.dpr as f32;
-
         // Update the window dimensions.
         if let Some(size) = update_pending.dimensions() {
-            // Ensure we have at least one column and row.
-            self.size_info.width = (size.width as f32).max(cell_width + 2. * padding_x);
-            self.size_info.height = (size.height as f32).max(cell_height + 2. * padding_y);
+            self.size_info.width = size.width as f32;
+            self.size_info.height = size.height as f32;
         }
 
-        // Distribute excess padding equally on all sides.
+        // Recalculate padding.
+        let (mut padding_x, mut padding_y) = scale_padding(config, self.size_info.dpr);
         if config.ui_config.window.dynamic_padding {
+            // Distribute excess padding equally on all sides.
             padding_x = dynamic_padding(padding_x, self.size_info.width, cell_width);
             padding_y = dynamic_padding(padding_y, self.size_info.height, cell_height);
         }
 
         self.size_info.padding_x = padding_x.floor() as f32;
         self.size_info.padding_y = padding_y.floor() as f32;
 
-        let mut pty_size = self.size_info;
-
-        // Subtract message bar lines from pty size.
-        if let Some(message) = message_buffer.message() {
-            let lines = message.text(&self.size_info).len();
-            pty_size.height -= pty_size.cell_height * lines as f32;
-        }
-
-        // Add an extra line for the current search regex.
-        if search_active {
-            pty_size.height -= pty_size.cell_height;
-        }
+        // Update number of column/lines in the viewport.
+        let message_bar_lines =
+            message_buffer.message().map(|m| m.text(&self.size_info).len()).unwrap_or(0);
+        let search_lines = if search_active { 1 } else { 0 };
+        self.size_info.update_dimensions(message_bar_lines + search_lines);
```

Comment:
I really don't like that `update_dimensions` API. It's **so** easy to screw things up, since you can clearly reach a state where width / cell_width != cols without any effort, or start updating just cols/lines. In general . The width/height of the pty that you'll be reporting via ioctl won't be correct either, but I'm not sure that it can confuse someone. though, if an application relies on that to position some overlay on top of terminal contents it could compute cell size incorrectly and do some weird stuff(I'm basically talking about some generic weird X11 image viewer that draws above terminal).

In general old compute style cols/lines made sense, and in a new impl you can always clamp returns from those functions?

Reply:
> I really don't like that update_dimensions API. It's so easy to screw things up, since you can clearly reach a state where width / cell_width != cols without any effort, or start updating just cols/lines.

How is this any different for any of the other fields of `SizeInfo`?

> The width/height of the pty that you'll be reporting via ioctl won't be correct either, but I'm not sure that it can confuse someone.

That depends on what these values are supposed to mean. Currently they're "unused" according to the documentation. The window size is certainly reported correctly, same as it did before. If it should report the text area size, it doesn't work just as much as it didn't work before. There should be zero change wrt ioctl size tbh.

> In general old compute style cols/lines made sense, and in a new impl you can always clamp returns from those functions?

How did it make sense to have multiple different reported terminal sizes, some including search/message bar and some excluding them, seemingly arbitrarily? How does it make sense to recompute a value every single time, when we know it never changes? The old behavior did a lot, but making sense it did not.

Reply:
>How is this any different for any of the other fields of SizeInfo?

They all cause re-computation of amount of lines/cols. Like if you change anything, everything will be properly updated.

>That depends on what these values are supposed to mean. Currently they're "unused" according to the documentation. The window size is certainly reported correctly, same as it did before. If it should report the text area size, it doesn't work just as much as it didn't work before. There should be zero change wrt ioctl size tbh.

it's not, you don't subtract message bar/search bar from it anymore.

>How did it make sense to have multiple different reported terminal sizes, some including search/message bar and some excluding them, seemingly arbitrarily? How does it make sense to recompute a value every single time, when we know it never changes? The old behavior did a lot, but making sense it did not.

We may made those fields as only getters and keep updat_dimensions. Direct access to those is bad.

Reply:
The point is that with old size info, changing each field result in proper cols/lines, however now you can change those fields directly, so you can miss use those, that's what I don't like about it.

Reply:
> They all cause re-computation of amount of lines/cols. Like if you change anything, everything will be properly updated.

They do, but anything changing in winit requires updating all of size info. It's never partially updated anyways.

> it's not, you don't subtract message bar/search bar from it anymore.

Yes I do?

> We may made those fields as only getters and keep updat_dimensions. Direct access to those is bad.

Sure, but it seems like all of this is just very dangerous to deal with. Like changing the width/height you shouldn't do without doing other stuff too (like updating dimensions). So should everything be done through one update method? But that doesn't work, since some things must be done separately (like DPR).

> The point is that with old size info, changing each field result in proper cols/lines, however now you can change those fields directly, so you can miss use those, that's what I don't like about it.

That is incorrect, since the message bar and search need to be subtracted. So you would still sometimes get incorrect results. And even worse now you don't actually know if you're allowed to call cols() and screen_lines() or if those are incorrect.